File: 소마/트리의지름.py
import sys

N = int(sys.stdin.readline().rstrip())
sys.setrecursionlimit(10**9)
graph = [[] for _ in range(N+1)]

for i in range(N):
  input = list(map(int,sys.stdin.readline().rstrip().split(' ')))
  length = len(input)//2-1
  
  for i in range(length):
      idx = 2*i+1
      graph[input[0]].append([input[idx],input[idx+1]])
    
distance = [-1 for _ in range(N+1)]
print(distance)
def dfs (node,w):
    for nextNode,nextWeight in graph[node]:
        if distance[nextNode]==-1:
            distance[nextNode] = nextWeight+w
            dfs(nextNode,nextWeight+w)

distance[1] = 0  
dfs(1,0)
startNode= distance.index(max(distance))
distance = [-1 for _ in range(N+1)]
distance[startNode] = 0  
dfs(startNode,0)

print(max(distance))

File: 소마/트리의지름.js
let input = `12
1 2 3
1 3 2
2 4 5
3 5 11
3 6 9
4 7 1
4 8 7
5 9 15
5 10 4
6 11 6
6 12 10`.split("\n");
const N = Number(input.shift());
const graph = {};

for (let i = 0; i < N - 1; i++) {
  [a, b, w] = input[i].split(" ").map(Number);
  graph[a] = graph[a] ? [...graph[a], [b, w]] : [[b, w]];
  graph[b] = graph[b] ? [...graph[b], [a, w]] : [[a, w]];
}

let max;
let leaf;
let visited;

const dfs = (node, cnt) => {
  if (cnt > max) {
    leaf = node;
    max = cnt;
  }

  for (let next of graph[node]) {
    const [nNode, nWeight] = next;
    if (visited[nNode]) continue;
    visited[nNode] = true;
    dfs(nNode, cnt + nWeight);
  }
};
max = -Infinity;
visited = Array.from({ length: N }, () => false);
dfs(1, 0);
max = -Infinity;
visited = Array.from({ length: N }, () => false);
dfs(leaf, 0);
console.log(max);
